File: nodejs/src/serialize.ts
import type { Address, Hex, Hop, Quote, SwapResult, TokenInfo } from "./types.js"

/**
 * `JSON.stringify` replacer that turns `bigint` values into base-10 strings.
 *
 *     JSON.stringify(obj, bigintReplacer)
 */
export function bigintReplacer(_key: string, value: unknown): unknown {
  return typeof value === "bigint" ? value.toString() : value
}

// ─── Quote ────────────────────────────────────────────────────────────────────

export interface SerializedHop extends Omit<Hop, "amountInWei" | "amountOutWei" | "minOutWei"> {
  amountInWei: string
  amountOutWei: string
  minOutWei: string
}

export interface SerializedQuote extends Omit<Quote, "amountInWei" | "amountOutWei" | "minOutWei" | "hops"> {
  amountInWei: string
  amountOutWei: string
  minOutWei: string
  hops: SerializedHop[]
  // network / maxHops / priceBase / dexs are JSON-safe primitives — inherited via Omit
}

function hopToJSON(h: Hop): SerializedHop {
  return { ...h, amountInWei: h.amountInWei.toString(), amountOutWei: h.amountOutWei.toString(), minOutWei: h.minOutWei.toString() }
}

function hopFromJSON(h: SerializedHop): Hop {
  return { ...h, amountInWei: BigInt(h.amountInWei), amountOutWei: BigInt(h.amountOutWei), minOutWei: BigInt(h.minOutWei) }
}

/** Converts a `Quote` to a JSON-safe object (all bigints become base-10 strings). */
export function quoteToJSON(q: Quote): SerializedQuote {
  return {
    ...q,
    amountInWei:  q.amountInWei.toString(),
    amountOutWei: q.amountOutWei.toString(),
    minOutWei:    q.minOutWei.toString(),
    hops:         q.hops.map(hopToJSON),
  }
}

/** Reverses {@link quoteToJSON}. Accepts either the typed shape or a generic object/string. */
export function quoteFromJSON(j: SerializedQuote | string): Quote {
  const o = typeof j === "string" ? JSON.parse(j) as SerializedQuote : j
  return {
    ...o,
    tokenIn:      o.tokenIn as Address,
    tokenOut:     o.tokenOut as Address,
    amountInWei:  BigInt(o.amountInWei),
    amountOutWei: BigInt(o.amountOutWei),
    minOutWei:    BigInt(o.minOutWei),
    steps:        o.steps as Hex,
    path:         (o.path ?? []) as Address[],
    hops:         (o.hops ?? []).map(hopFromJSON),
  }
}

// ─── SwapResult ───────────────────────────────────────────────────────────────

export interface SerializedSwapResult extends Omit<SwapResult, "amountIn" | "amountOut" | "blockNumber" | "gasUsed" | "effectiveGasPrice" | "feeWei"> {
  amountIn: string
  amountOut: string
  blockNumber: string
  gasUsed: string
  effectiveGasPrice: string
  feeWei: string
}

export function swapResultToJSON(r: SwapResult): SerializedSwapResult {
  return {
    ...r,
    amountIn:    r.amountIn.toString(),
    amountOut:   r.amountOut.toString(),
    blockNumber: r.blockNumber.toString(),
    gasUsed:     r.gasUsed.toString(),
    effectiveGasPrice: r.effectiveGasPrice.toString(),
    feeWei:      r.feeWei.toString(),
  }
}

export function swapResultFromJSON(j: SerializedSwapResult | string): SwapResult {
  const o = typeof j === "string" ? JSON.parse(j) as SerializedSwapResult : j
  return {
    ...o,
    txHash:      o.txHash as Hex,
    tokenIn:     o.tokenIn as Address,
    tokenOut:    o.tokenOut as Address,
    amountIn:    BigInt(o.amountIn),
    amountOut:   BigInt(o.amountOut),
    blockNumber: BigInt(o.blockNumber),
    gasUsed:     BigInt(o.gasUsed),
    effectiveGasPrice: BigInt(o.effectiveGasPrice ?? "0"),
    feeWei:      BigInt(o.feeWei ?? "0"),
  }
}

// ─── TokenInfo ───────────────────────────────────────────────────────────────

export interface SerializedTokenInfo extends Omit<TokenInfo, "balance" | "allowance"> {
  balance?: string
  allowance?: string
}

export function tokenInfoToJSON(t: TokenInfo): SerializedTokenInfo {
  return {
    ...t,
    balance:   t.balance   !== undefined ? t.balance.toString()   : undefined,
    allowance: t.allowance !== undefined ? t.allowance.toString() : undefined,
  }
}

export function tokenInfoFromJSON(j: SerializedTokenInfo | string): TokenInfo {
  const o = typeof j === "string" ? JSON.parse(j) as SerializedTokenInfo : j
  return {
    ...o,
    address:   o.address as Address,
    owner:     o.owner as Address | undefined,
    balance:   o.balance   !== undefined ? BigInt(o.balance)   : undefined,
    allowance: o.allowance !== undefined ? BigInt(o.allowance) : undefined,
  }
}

File: go/serialize.go
package afi

import (
	"encoding/json"
	"fmt"
	"math/big"

	"github.com/ethereum/go-ethereum/common"
)

// JSON serialization helpers.
//
// Default *big.Int marshaling writes raw decimal numbers, which lose precision
// once they exceed JavaScript's safe integer range. The Marshal/Unmarshal pairs
// here serialize *big.Int as base-10 strings so the payload round-trips cleanly
// across language boundaries.

type hopJSON struct {
	TokenIn       common.Address `json:"tokenIn"`
	TokenOut      common.Address `json:"tokenOut"`
	AmountIn      string         `json:"amountIn"`
	AmountOut     string         `json:"amountOut"`
	MinOut        string         `json:"minOut"`
	AmountInWei   string         `json:"amountInWei"`
	AmountOutWei  string         `json:"amountOutWei"`
	MinOutWei     string         `json:"minOutWei"`
	TokenInPrice  string         `json:"tokenInPrice"`
	TokenOutPrice string         `json:"tokenOutPrice"`
	Slippage      float64        `json:"slippage"`
	Type          string         `json:"type"`
	Kind          string         `json:"kind"`
	RouteID       int            `json:"routeId"`
	Weight        float64        `json:"weight"`
}

type quoteJSON struct {
	TokenIn           common.Address   `json:"tokenIn"`
	TokenOut          common.Address   `json:"tokenOut"`
	AmountIn          string           `json:"amountIn"`
	AmountOut         string           `json:"amountOut"`
	MinOut            string           `json:"minOut"`
	AmountInWei       string           `json:"amountInWei"`
	AmountOutWei      string           `json:"amountOutWei"`
	MinOutWei         string           `json:"minOutWei"`
	Steps             string           `json:"steps"`
	Path              []common.Address `json:"path"`
	Hops              []hopJSON        `json:"hops"`
	Slippage          float64          `json:"slippage"`
	FeeBps            uint16           `json:"feeBps"`
	TokenInPrice      string           `json:"tokenInPrice"`
	TokenOutPrice     string           `json:"tokenOutPrice"`
	TokenInBasePrice  string           `json:"tokenInBasePrice,omitempty"`
	TokenOutBasePrice string           `json:"tokenOutBasePrice,omitempty"`
	CreatedAt         int64            `json:"createdAt"`
	Network           Network          `json:"network"`
	MaxHops           int              `json:"maxHops"`
	PriceBase         string           `json:"priceBase,omitempty"`
	Dexs              []Dex            `json:"dexs,omitempty"`
}

func bigIntString(b *big.Int) string {
	if b == nil {
		return "0"
	}
	return b.String()
}

func parseBigInt(s string) (*big.Int, error) {
	b, ok := new(big.Int).SetString(s, 10)
	if !ok {
		return nil, fmt.Errorf("invalid bigint string: %q", s)
	}
	return b, nil
}

// MarshalJSON serializes Quote with *big.Int fields as base-10 strings.
func (q *Quote) MarshalJSON() ([]byte, error) {
	hops := make([]hopJSON, len(q.Hops))
	for i, h := range q.Hops {
		hops[i] = hopJSON{
			TokenIn: h.TokenIn, TokenOut: h.TokenOut,
			AmountIn: h.AmountIn, AmountOut: h.AmountOut, MinOut: h.MinOut,
			AmountInWei: bigIntString(h.AmountInWei), AmountOutWei: bigIntString(h.AmountOutWei), MinOutWei: bigIntString(h.MinOutWei),
			TokenInPrice: h.TokenInPrice, TokenOutPrice: h.TokenOutPrice,
			Slippage: h.Slippage, Type: h.Type, Kind: h.Kind, RouteID: h.RouteID, Weight: h.Weight,
		}
	}
	return json.Marshal(quoteJSON{
		TokenIn: q.TokenIn, TokenOut: q.TokenOut,
		AmountIn: q.AmountIn, AmountOut: q.AmountOut, MinOut: q.MinOut,
		AmountInWei: bigIntString(q.AmountInWei), AmountOutWei: bigIntString(q.AmountOutWei), MinOutWei: bigIntString(q.MinOutWei),
		Steps: "0x" + common.Bytes2Hex(q.Steps), Path: q.Path, Hops: hops,
		Slippage: q.Slippage, FeeBps: q.FeeBps,
		TokenInPrice: q.TokenInPrice, TokenOutPrice: q.TokenOutPrice,
		TokenInBasePrice: q.TokenInBasePrice, TokenOutBasePrice: q.TokenOutBasePrice,
		CreatedAt: q.CreatedAt,
		Network:   q.Network, MaxHops: q.MaxHops, PriceBase: q.PriceBase, Dexs: q.Dexs,
	})
}

// UnmarshalJSON reverses MarshalJSON.
func (q *Quote) UnmarshalJSON(data []byte) error {
	var j quoteJSON
	if err := json.Unmarshal(data, &j); err != nil {
		return err
	}
	amountInWei, err := parseBigInt(j.AmountInWei)
	if err != nil {
		return fmt.Errorf("amountInWei: %w", err)
	}
	amountOutWei, err := parseBigInt(j.AmountOutWei)
	if err != nil {
		return fmt.Errorf("amountOutWei: %w", err)
	}
	minOutWei, err := parseBigInt(j.MinOutWei)
	if err != nil {
		return fmt.Errorf("minOutWei: %w", err)
	}
	hops := make([]Hop, len(j.Hops))
	for i, h := range j.Hops {
		a, errA := parseBigInt(h.AmountInWei)
		b, errB := parseBigInt(h.AmountOutWei)
		m, errM := parseBigInt(h.MinOutWei)
		if errA != nil || errB != nil || errM != nil {
			return fmt.Errorf("hop %d bigint: %v/%v/%v", i, errA, errB, errM)
		}
		hops[i] = Hop{
			TokenIn: h.TokenIn, TokenOut: h.TokenOut,
			AmountIn: h.AmountIn, AmountOut: h.AmountOut, MinOut: h.MinOut,
			AmountInWei: a, AmountOutWei: b, MinOutWei: m,
			TokenInPrice: h.TokenInPrice, TokenOutPrice: h.TokenOutPrice,
			Slippage: h.Slippage, Type: h.Type, Kind: h.Kind, RouteID: h.RouteID, Weight: h.Weight,
		}
	}
	*q = Quote{
		TokenIn: j.TokenIn, TokenOut: j.TokenOut,
		AmountIn: j.AmountIn, AmountOut: j.AmountOut, MinOut: j.MinOut,
		AmountInWei: amountInWei, AmountOutWei: amountOutWei, MinOutWei: minOutWei,
		Steps: common.FromHex(j.Steps), Path: j.Path, Hops: hops,
		Slippage: j.Slippage, FeeBps: j.FeeBps,
		TokenInPrice: j.TokenInPrice, TokenOutPrice: j.TokenOutPrice,
		TokenInBasePrice: j.TokenInBasePrice, TokenOutBasePrice: j.TokenOutBasePrice,
		CreatedAt: j.CreatedAt,
		Network:   j.Network, MaxHops: j.MaxHops, PriceBase: j.PriceBase, Dexs: j.Dexs,
	}
	return nil
}

type swapResultJSON struct {
	TxHash            common.Hash    `json:"txHash"`
	BlockNumber       uint64         `json:"blockNumber"`
	AmountIn          string         `json:"amountIn"`
	AmountOut         string         `json:"amountOut"`
	TokenIn           common.Address `json:"tokenIn"`
	TokenOut          common.Address `json:"tokenOut"`
	GasUsed           uint64         `json:"gasUsed"`
	EffectiveGasPrice string         `json:"effectiveGasPrice,omitempty"`
	FeeWei            string         `json:"feeWei,omitempty"`
	FeeETH            string         `json:"feeEth,omitempty"`
}

func (r *SwapResult) MarshalJSON() ([]byte, error) {
	return json.Marshal(swapResultJSON{
		TxHash: r.TxHash, BlockNumber: r.BlockNumber,
		AmountIn: bigIntString(r.AmountIn), AmountOut: bigIntString(r.AmountOut),
		TokenIn: r.TokenIn, TokenOut: r.TokenOut, GasUsed: r.GasUsed,
		EffectiveGasPrice: bigIntString(r.EffectiveGasPrice),
		FeeWei:            bigIntString(r.FeeWei),
		FeeETH:            r.FeeETH,
	})
}

func (r *SwapResult) UnmarshalJSON(data []byte) error {
	var j swapResultJSON
	if err := json.Unmarshal(data, &j); err != nil {
		return err
	}
	amountIn, err := parseBigInt(j.AmountIn)
	if err != nil {
		return fmt.Errorf("amountIn: %w", err)
	}
	amountOut, err := parseBigInt(j.AmountOut)
	if err != nil {
		return fmt.Errorf("amountOut: %w", err)
	}
	out := SwapResult{
		TxHash: j.TxHash, BlockNumber: j.BlockNumber,
		AmountIn: amountIn, AmountOut: amountOut,
		TokenIn: j.TokenIn, TokenOut: j.TokenOut, GasUsed: j.GasUsed,
		FeeETH: j.FeeETH,
	}
	if j.EffectiveGasPrice != "" {
		if v, err := parseBigInt(j.EffectiveGasPrice); err == nil {
			out.EffectiveGasPrice = v
		}
	}
	if j.FeeWei != "" {
		if v, err := parseBigInt(j.FeeWei); err == nil {
			out.FeeWei = v
		}
	}
	*r = out
	return nil
}

type tokenInfoJSON struct {
	Address   common.Address `json:"address"`
	Symbol    string         `json:"symbol"`
	Name      string         `json:"name"`
	Decimals  uint8          `json:"decimals"`
	Owner     common.Address `json:"owner,omitempty"`
	Balance   string         `json:"balance,omitempty"`
	Allowance string         `json:"allowance,omitempty"`
}

func (t *TokenInfo) MarshalJSON() ([]byte, error) {
	j := tokenInfoJSON{
		Address: t.Address, Symbol: t.Symbol, Name: t.Name, Decimals: t.Decimals,
		Owner: t.Owner,
	}
	if t.Balance != nil {
		j.Balance = t.Balance.String()
	}
	if t.Allowance != nil {
		j.Allowance = t.Allowance.String()
	}
	return json.Marshal(j)
}

func (t *TokenInfo) UnmarshalJSON(data []byte) error {
	var j tokenInfoJSON
	if err := json.Unmarshal(data, &j); err != nil {
		return err
	}
	out := TokenInfo{
		Address: j.Address, Symbol: j.Symbol, Name: j.Name, Decimals: j.Decimals,
		Owner: j.Owner,
	}
	if j.Balance != "" {
		b, err := parseBigInt(j.Balance)
		if err != nil {
			return fmt.Errorf("balance: %w", err)
		}
		out.Balance = b
	}
	if j.Allowance != "" {
		a, err := parseBigInt(j.Allowance)
		if err != nil {
			return fmt.Errorf("allowance: %w", err)
		}
		out.Allowance = a
	}
	*t = out
	return nil
}
